src/or/onion.c

```diff
@@ -1339,3 +1342,466 @@ extended_cell_format(uint8_t *command_out, uint16_t *len_out,
+    + 1; // for the uint8_t trunnel_error_code_
+}
+
+/**
+ * Check that the encoded length of an incoming CREATE2V or CREATED2V cell is
+ * less than the maximum allowed.
+ *
+ * Returns true if it's okay, and false otherwise.
+ */
+static bool
+create2v_cell_check_encoded_length(const size_t len)
+{
+  const uint16_t max_len = create2v_cell_calculate_maximum_total();
+
+  /* If the incoming buffer is reportedly way larger than that which would fill
+   * the allowable space for hdata and the ignored padding, then fail parsing
+   * early.
+   */
+  if (len > max_len) {
+    log_info(LD_OR, "Received a suspicious CREATE(D)2V cell whose reported "
+             "length (%zd bytes) was greater that the maximum allowed (%hd "
+             "bytes). This is either an attempted DoS, or there exist "
+             "newer/other onion routers speaking a circuit handshake we "
+             "don't know about.", len, max_len);
+    return false;
+  }
+
+  if (len == 0) {
+    log_info(LD_OR, "Received a strange CREATE(D)2V cell whose reported "
+             "length was 0 bytes.");
+    return false;
+  }
+
+  return true;
+}
+
+/**
+ * Parse the CREATE2V payload at <b>p</b>, which could be up to <b>p_len</b>
+ * bytes long, and use it to fill the fields of <b>cell_out->body</b>.
+ *
+ * Note that part of the body of an EXTEND2V cell is a CREATE2V payload, so
+ * this function is also used for parsing those.
+ *
+ * Returns true if parsing was successful, and false otherwise.
+ */
+static bool
+parse_create2v_payload(create2v_cell_t *cell_out,
+                       const uint8_t *p,
+                       const size_t p_len)
+{
+  create2v_cell_body_t *body;
+
+  /* Check the length before even attempting to parse it. */
+  if (!create2v_cell_check_encoded_length(p_len)) {
+    return false;
+  }
+  create2v_cell_body_parse(&body, p, p_len);
+  cell_out->body = body;
+
+  return create2v_cell_check(cell_out, true);
+}
+
+/**
+ * Parse a CREATE2V cell from <b>cell_in</b> into <b>cell_out</b>.
+ *
+ * Returns true if parsing was successful, and false otherwise.
+ */
+bool
+create2v_cell_parse(create2v_cell_t *cell_out, const var_cell_t *cell_in)
+{
+  switch (cell_in->headers.command) {
+  case CELL_CREATE2V:
+    return parse_create2v_payload(cell_out,
+                                  cell_in->payload,
+                                  cell_in->payload_len);
+  default:
+    return false;
+  }
+}
+
+/**
+ * Run various checks to ensure that a create2v_cell_t, <b>cell</b>, is
+ * well-formed.
+ *
+ * If <b>unknown_ok</b> is true, then handshake types we don't know about yet
+ * are okay.
+ *
+ * In the future, checks such as verifying that the handshake data (for a
+ * particular handshake) should go here.
+ *
+ * Returns true if the <b>cell</b> is okay, and false otherwise.
+ */
+bool
+create2v_cell_check(const create2v_cell_t *cell, bool unknown_ok)
+{
+  const char *not_okay;
+
+  not_okay = create2v_cell_body_check(cell->body);
```

Comment:
I think this trunnel function can't actually fail on a parsed body ... only on a body we're thinking of parsing.